### PR TITLE
Sixth dan

### PR DESCRIPTION
**Pull request by gigjozsa**
Again carate and Jenkins. In this version
- caratekit is modified to allow the home directory to be redirected to any location
- caratekit now allows to preserve the home directory (and hence all caches inside home)
- the home directory is redirected to a place in the directory tree that is shared with all pull request builds.

After running several times (filling up all caches) a double test takes an hour on ray.

### CODE REVIEW

**Comment by gigjozsa** [comment]
retest this please

**Comment by gigjozsa** [comment]
retest this please

**Comment by gigjozsa** [comment]
retest this please

**Comment by gigjozsa** [comment]
@bennahugo @o-smirnov I have no clue why this is failing with an enigmatic error in Jenkins:
```/var/lib/jenkins/jobs/PR_meerkathi/workspace/103/test-output/d9dbd04c7429a7f97c398cf6407422c3860c68a8/caracal_venv/bin/stimela: line 3: import: command not found
```
Has Stimela been changed? Is the shebang suddenly missing?

**Comment by o-smirnov** [comment]
Hmmm, works for me on the VM. Shebang is in place. @gigjozsa has anything changed about how you invoke it?
```
$ virtualenv -p python3 test
Already using interpreter /usr/bin/python3
Using base prefix '/usr'
New python executable in /home/oms/test/bin/python3
Also creating executable in /home/oms/test/bin/python
Installing setuptools, pkg_resources, pip, wheel...done.
$ . test/bin/activate
(test) $ pip install stimela
Collecting stimela
  Downloading stimela-1.6.2.tar.gz (186 kB)
     |████████████████████████████████| 186 kB 1.2 MB/s 
Collecting pyyaml
  Downloading PyYAML-5.3.1.tar.gz (269 kB)
     |████████████████████████████████| 269 kB 5.0 MB/s 
Collecting nose>=1.3.7
  Downloading nose-1.3.7-py3-none-any.whl (154 kB)
     |████████████████████████████████| 154 kB 4.1 MB/s 
Collecting future-fstrings
  Downloading future_fstrings-1.2.0-py2.py3-none-any.whl (6.1 kB)
Building wheels for collected packages: stimela, pyyaml
  Building wheel for stimela (setup.py) ... done
  Created wheel for stimela: filename=stimela-1.6.2-py3-none-any.whl size=289702 sha256=3647def99e33f7b94e292fc41f1b4145581900bcf57a397d0e3971aec7d36bdc
  Stored in directory: /home/oms/.cache/pip/wheels/a6/ac/0c/05c7a137d4717f174203d0cfff4ec71b501e26489fa7b626df
  Building wheel for pyyaml (setup.py) ... done
  Created wheel for pyyaml: filename=PyYAML-5.3.1-cp36-cp36m-linux_x86_64.whl size=44621 sha256=9595e287c686b14201d69a36949d35ccbad34564d2c7de6ac1f08b5ee076c55a
  Stored in directory: /home/oms/.cache/pip/wheels/e5/9d/ad/2ee53cf262cba1ffd8afe1487eef788ea3f260b7e6232a80fc
Successfully built stimela pyyaml
Installing collected packages: pyyaml, nose, future-fstrings, stimela
Successfully installed future-fstrings-1.2.0 nose-1.3.7 pyyaml-5.3.1 stimela-1.6.2
(test) $ stimela
usage: stimela [-h] [-v] [command [options] [command [options] ...]]
```

**Comment by gigjozsa** [comment]
@o-smirnov I made a few changes but the test works fine on e.g. ray for me. Short and sweet. This is what the script does did (last line is a trap on error):
```
workspace=/var/lib/jenkins/jobs/PR_meerkathi/workspace/103/test-output
test_data_dir=/var/lib/jenkins/jobs/PR_meerkathi/workspace/103/test-output/pull_request_data/
caracal_test_id=d9dbd04c7429a7f97c398cf6407422c3860c68a8
local_caracal=/var/lib/jenkins/jobs/PR_meerkathi/workspace/103/projects/caracal/
workspace_root=${workspace}/${caracal_test_id}
cvirtualenv=${workspace_root}/caracal_venv
HOME_OLD=${HOME}

mkdir -p ${workspace_root}
export HOME=/var/lib/jenkins/jobs/PR_meerkathi/workspace/home
cd ${HOME}
rm -rf ${cvirtualenv}
python3 -m venv ${cvirtualenv}
rm -rf ${cvirtualenv}
virtualenv -p python3 ${cvirtualenv}
. ${cvirtualenv}/bin/activate
export PYTHONPATH=''
pip install pip setuptools wheel -U
rm -rf ${workspace_root}/caracal
cd ${workspace_root}
cp -r ${local_caracal} ${workspace_root}/
pip install -U --force-reinstall ${workspace_root}/caracal
stimela clean -ac
export HOME=${HOME_OLD}
```

**Comment by o-smirnov** [comment]
Add ``which stimela; head `which stimela` `` before you invoke it? I mean this is all clearly impossible by definition, so let's keep digging...

**Comment by gigjozsa** [comment]
```
/var/lib/jenkins/jobs/PR_meerkathi/workspace/104/test-output/d805fc22d446c873b22a6724db3adda060ca0e19/caracal_venv/bin/stimela
#!/var/lib/jenkins/jobs/PR_meerkathi/workspace/104/test-output/d805fc22d446c873b22a6724db3adda060ca0e19/caracal_venv/bin/python3

import os
import sys

import stimela
import stimela.main as main

# Ensure that logging infrastructure exists
```

**Comment by gigjozsa** [comment]
First line from the above is from the which stimela. Seems correct.

**Comment by o-smirnov** [comment]
I dunno @gigjozsa, somehow you have contrived to create a deeply broken virtualenv. Even a basic script executed outside the virtualenv:

```
#!/var/lib/jenkins/jobs/PR_meerkathi/workspace/104/test-output/d805fc22d446c873b22a6724db3adda060ca0e19/caracal_venv/bin/python3
import sys
```

uses shell and not python. It's like that python3 executable substitutes the shell for itself for some reason. I don't know how this is possible, but this is clearly happening.

If I create a python3 virtualenv by hand, everything works fine.

Maybe it's the monkeying around with ``HOME`` that you're doing that somehow confuses it?

**Comment by gigjozsa** [comment]
Yes. On the other hand this happens only here. No problemo doing the same stuff elsewhere.

**Comment by gigjozsa** [comment]
retest this please

**Comment by gigjozsa** [comment]
What I did was to activate "Discard new builds" in the setup section of Jenkins and then switched it off again. Cannot imagine that that caused the above stuff.

**Comment by gigjozsa** [comment]
@bennahugo @o-smirnov there is a button in the Jekins page "wipe out current workspace". This should enable a completely new start, right? Can I use it?

**Comment by gigjozsa** [comment]
OK, I suggest to give up on the Jenkins tests. This runs smoothly as a very smooth thing on any machine that I am trying it out on, but there seem to be some incompatibilities that I do not understand on the Jenkins VM. The only thing that I suggest is to wipe out the whole workspace and see if it runs after that. Please advise. 

Apart from that I will update the test instructions with a version where people can re-use pull folders and that's it.

**Comment by o-smirnov** [comment]
OK so @gigjozsa started build 121. Let's let it run until the failure, then @bennahugo and me can pick over the bones.

**Comment by gigjozsa** [comment]
Finished, same error.

File: caratekit.sh [hunk 1]
```diff
@@ -70,6 +70,7 @@ IA=5
 endimessage=""
 ALTERNATIVE_TEST_DATA_IDS=( "1524929477" "1524947605" "1532022061" )
 SHORT_TEST_DATA_IDS=( "1532022061_subset" )
+SPHEVERSION="1.1.0"
```
Comment (o-smirnov) [inline caratekit.sh:73]: Are we versioning @SpheMakh now? Just curious.

Comment (paoloserra) [reply caratekit.sh:73]: we totally should!

**Comment by o-smirnov** [comment]
> As it stands we can now have 999 999 999 999 999 999 999 999 999 999 999 999 999 tests before the problem reoccurs.

In other words, if we have another busy-week schedule-slip into a busy-month, we should be worried.

File: caratekit.sh (hunk 1 above)
Comment (gigjozsa) [reply caratekit.sh:73]: @SpheMakh is improving all the time, so yes, of course.

**Pull request merged**